scripts/builtin/gaussianClassifier.dml

```diff
@@ -0,0 +1,127 @@
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+#
+# Computes the parameters needed for Gaussian Classification.
+# Thus it computes the following per class: the prior probability,
+# the inverse covariance matrix, the mean per feature and the determinant
+# of the covariance matrix. Furthermore (if not explicitely defined), it
+# adds some small smoothing value along the variances, to prevent
+# numerical errors / instabilities.
+#
+#
+# INPUT PARAMETERS:
+# -------------------------------------------------------------------------------------------------
+# NAME           TYPE               DEFAULT  MEANING
+# -------------------------------------------------------------------------------------------------
+# D              Matrix[Double]     ---      Input matrix (training set)
+# C              Matrix[Double]     ---      Target vector
+# varSmoothing   Double             1e-9     Smoothing factor for variances
+# verbose        Boolean            TRUE     Print accuracy of the training set
+# ---------------------------------------------------------------------------------------------
+# OUTPUT:
+# ---------------------------------------------------------------------------------------------
+# NAME                  TYPE             DEFAULT  MEANING
+# ---------------------------------------------------------------------------------------------
+# classPriors           Matrix[Double]   ---      Vector storing the class prior probabilities
+# classMeans            Matrix[Double]   ---      Matrix storing the means of the classes
+# classInvCovariances   List[Unknown]    ---      List of inverse covariance matrices
+# determinants          Matrix[Double]   ---      Vector storing the determinants of the classes
+# ---------------------------------------------------------------------------------------------
+#
+
+
+m_gaussianClassifier = function(Matrix[Double] D, Matrix[Double] C, Double varSmoothing=1e-9, Boolean verbose = TRUE)
+  return (Matrix[Double] classPriors, Matrix[Double] classMeans,
+  List[Unknown] classInvCovariances, Matrix[Double] determinants)
+{
+  #Retrieve number of samples, classes and features
+  nSamples = nrow(D)
+  nClasses = max(C)
+  nFeats = ncol(D)
+
+  #Set varSmoothing (to prevent numerical errors)
+  varSmoothing = 1e-9
+
+  #Compute means, variances and priors
+  classCounts = aggregate(target=C, groups=C, fn="count", ngroups=as.integer(nClasses));
+  classMeans = aggregate(target=D, groups=C, fn="mean", ngroups=as.integer(nClasses));
+  classVars = aggregate(target=D, groups=C, fn="variance", ngroups=as.integer(nClasses));
+  classPriors = classCounts / nSamples
+
+  smoothedVar = diag(matrix(1.0, rows=nFeats, cols=1)) * max(classVars) * varSmoothing
+
+  classInvCovariances = list()
+  determinants = matrix(0, rows=nClasses, cols=1)
+
+  #Compute determinants and inverseCovariances
+  for (class in 1:nClasses)
```

Comment:
consider if it can be parallelized, if so use a parfor

Reply:
I pushed a new commit. Unfortunately parfor in this case was not possible, since I append the results to a list, which can not be parallelized.